src/main/java/org/springframework/retry/backoff/ExponentialBackOffPolicy.java

```diff
@@ -289,15 +289,25 @@ public synchronized long getSleepAndIncrement() {
 		}
 
 		protected long getNextInterval() {
-			return (long) (this.interval * getMultiplier());
+			if (this.initialIntervalSupplier == null || this.interval != DEFAULT_INITIAL_INTERVAL) {
+				return (long) (this.interval * getMultiplier());
+			}
+			else {
+				return (long) (this.initialIntervalSupplier.get() * getMultiplier());
+			}
 		}
 
 		public double getMultiplier() {
 			return this.multiplierSupplier != null ? this.multiplierSupplier.get() : this.multiplier;
 		}
 
 		public long getInterval() {
-			return this.intervalSupplier != null ? this.intervalSupplier.get() : this.interval;
+			if (initialIntervalSupplier == null || this.interval != DEFAULT_INITIAL_INTERVAL) {
```

Comment:
Why are we changing the logic to make the `interval` superior to the `intervalSupplier`?

Perhaps a description of exactly what you are trying to "fix" in the PR description would have been helpful.

Reply:
Sorry, I am still missing something; with the existing code...

```
return this.intervalSupplier != null ? this.intervalSupplier.get() : this.interval;
```

...the supplier, if present, takes precedence over the `interval`, regardless of whether it is the default value or not.

Also, it is not at all clear to me why we are testing against `DEFAULT_INITIAL_INTERVAL` - this value could be reached at some point after the initial delay - e.g. if the initial interval returned by the supplier is `DEFAULT_INITIAL_INTERVAL / 2` and the `multiplier` is 2.0.

Perhaps you could provide a test case that fails with the existing code, it would help me understand what you are trying to fix.

Reply:
Sorry didn't understand line of code your referring, is it before my change?

You can try test I added in this PR it is failing without my change.
```
	@Test
	public void testMultiBackOffWithInitialDelaySupplier() {
		ExponentialBackOffPolicy strategy = new ExponentialBackOffPolicy();
		long seed = 40;
		double multiplier = 1.2;
		strategy.initialIntervalSupplier(() -> 40L);
		strategy.setMultiplier(multiplier);
		strategy.setSleeper(sleeper);
		BackOffContext context = strategy.start(null);
		for (int x = 0; x < 5; x++) {
			strategy.backOff(context);
			assertThat(sleeper.getLastBackOff()).isEqualTo(seed);
			seed *= multiplier;
		}
	}
```

Reply:
The problem, @garyrussell , that `getSleepAndIncrement()` when calling `this.interval = getNextInterval();` does not take into account `intervalSupplier`.
And this one we need to call there only once: for initial `interval` value since `this.interval` is not set in case of `intervalSupplier`.

Reply:
Ok; I see the problem now, thanks; but I think this is a much simpler fix (and doesn't require any tests against the constant which, as I stated, may cause undesirable side effects).

```java
public long getInterval() {
	if (this.intervalSupplier != null) {
		this.interval = this.intervalSupplier.get();
		this.intervalSupplier = null;
	}
	return this.interval;
}
```

Reply:
... and I guess we need to set its result into `this.interval` since `getNextInterval()` is going to use that one only.

Reply:
Yes, 3rd line in the above code.

Reply:
updated according @garyrussell suggestion.